rfcs/20220221-3d_model_support.md

```diff
@@ -0,0 +1,181 @@
+
+## Some other considerations
+
+* It would be ideal to have this under a feature flag in the API. We could perform deployments and testing on our staging setup with the flag enabled, but still make other deployments to production with the flag disabled. We should be able to run the API for images & audio successfully with the feature flag disabled AND without having run the migrations.
+* As we work through this on both the catalog and the API, we should begin compiling information for a document describing the process for adding a **new media type**. This will be useful for future media types we are sure to add!
+
+## Prior work
+
+We may be able to use previous work regarding the addition of audio to the catalog/API as a guide. These updates required a fair amount of generalization, which we may not need to do a second time around.
+
+WordPress/openverse#19
+
+## Technical implementation
+
+### 1. Add 3D models to the catalog
+
+- [ ] Define necessary fields & schema for `model_3d` table & associated data
+- [ ] Add DDL to local Postgres image: https://github.com/WordPress/openverse-catalog/tree/main/docker/local_postgres
+- [ ] Create a popularity calculation view for `model_3d`
+- [ ] Add the database columns to `db_columns.py`: https://github.com/WordPress/openverse-catalog/blob/main/openverse_catalog/dags/common/storage/db_columns.py
+- [ ] Add the TSV columns to `tsv_columns.py`: https://github.com/WordPress/openverse-catalog/blob/main/openverse_catalog/dags/common/storage/tsv_columns.py
+- [ ] Create a `Model3DStore` subclass of [`MediaStore`](https://github.com/WordPress/openverse-catalog/blob/9538f38401e5428b32dc14c5d5ba7ef50af87a96/openverse_catalog/dags/common/storage/media.py#L39): 
+- [ ] Update the [new provider API script template](https://github.com/WordPress/openverse-catalog/tree/main/openverse_catalog/templates) with the `model_3d` provider
+- [ ] Create one/several provider API scripts & associated DAGs
+
+### 2. Add 3D models to the API
+_Note: Borrowed heavily from WordPress/openverse#19_
+- [ ] Django models for `model_3d`
+- [ ] Add DRF serializers for `model_3d`
+- [ ] Add DRF endpoints for `model_3d`
+- [ ] Make the thumbnail server work with model thumbnails
+
+### 3. Add 3D models to the data refresh
+- [ ] Add Elasticsearch `model_3d` models
+- [ ] Ensure the data refresh works for 3D models
+- [ ] Add 3D model sample data into the sample data CSV for testing
+
+
+## Open Questions
+
+* What data is important to store in the catalog? Some of this is encapsulated in the proposed schema, but there may be other important attributes beyond those listed.
+* Will it be important to differentiate the "type" of 3D model that a particular record refers to? (E.g. 3D printer model, game asset, artwork, etc.)
+* Many model sites come with asset groups, should we have a table similar to `AudioSets` for these cases?
```

Comment:
Aside: It appears that AudioSets is actually just a json blob on the Audio row.

Why is that not a foreign key I wonder?

Reply:
Hmm, I don't think it is? It looks like it's a separate table to me: https://github.com/WordPress/openverse-api/blob/c5daff0a090883d2232394872293796734689640/api/catalog/api/models/audio.py#L186

```
openledger> \dt audio*;
+--------+------------------+-------+--------+
| Schema | Name             | Type  | Owner  |
|--------+------------------+-------+--------|
| public | audio            | table | deploy |
| public | audiolist        | table | deploy |
| public | audiolist_audios | table | deploy |
| public | audioset         | table | deploy |
+--------+------------------+-------+--------+
SELECT 4
Time: 0.087s
openledger> \d audioset;
+---------------------+--------------------------+-------------------------------------+
| Column              | Type                     | Modifiers                           |
|---------------------+--------------------------+-------------------------------------|
| id                  | integer                  |  not null                           |
| created_on          | timestamp with time zone |  not null default CURRENT_TIMESTAMP |
| updated_on          | timestamp with time zone |  not null default CURRENT_TIMESTAMP |
| foreign_identifier  | character varying(1000)  |                                     |
| title               | character varying(2000)  |                                     |
| foreign_landing_url | character varying(1000)  |                                     |
| creator             | character varying(2000)  |                                     |
| creator_url         | character varying(2000)  |                                     |
| url                 | character varying(1000)  |                                     |
| filesize            | integer                  |                                     |
| filetype            | character varying(80)    |                                     |
| thumbnail           | character varying(1000)  |                                     |
| provider            | character varying(80)    |                                     |
+---------------------+--------------------------+-------------------------------------+
Indexes:
    "api_audioset_pkey" PRIMARY KEY, btree (id)
    "api_audioset_url_key" UNIQUE, btree (url)
    "unique_foreign_identifier_provider" UNIQUE, btree (foreign_identifier, provider)
    "api_audioset_foreign_identifier_28c1db59" btree (foreign_identifier)
    "api_audioset_foreign_identifier_28c1db59_like" btree (foreign_identifier varchar_pattern_ops)
    "api_audioset_url_200f9fb4_like" btree (url varchar_pattern_ops)
    "audioset_provider_cc061f30" btree (provider)
    "audioset_provider_cc061f30_like" btree (provider varchar_pattern_ops)

Time: 0.551s
```

I could be misinterpreting your question :slightly_smiling_face:

Reply:
I was going off of the DDL in the catalog, but maybe it gets turned into a real table when transferred to the API DB?

https://github.com/WordPress/openverse-catalog/blob/0c5b43fbab8cba491817ebb647b77e17114a7353/docker/local_postgres/0007_openledger_audio_view.sql#L90-L103

https://github.com/WordPress/openverse-catalog/blob/6c172033e42a91bcd8f9bf78fd6b933a70bd88bf/archive/openledger_sql/03_create_audio_schema.sql#L33-L35

Reply:
Ah yea, that's exactly what happens!

https://github.com/WordPress/openverse-api/blob/1d186bcf9d1df29d917bd79847840f82cf20e0ce/ingestion_server/ingestion_server/tasks.py#L115

I can elaborate a bit more on this point to denote the distinction (since it'll likely be a similar setup).

Reply:
Is that more performant for ingestion because it eliminates a FK? It seems counter-intuitive that storing _more_ data overall would be better (especially as jsonb).

Reply:
I'm almost certain that's the reason that's the reason why this is in the primary table as `jsonb` and pulled out into a view, maybe @obulat and @krysal might know more?